File: src/main/java/com/likelionknu/applyserver/auth/service/AuthService.java
package com.likelionknu.applyserver.auth.service;

import com.likelionknu.applyserver.auth.data.dto.response.TokenResponseDto;
import com.likelionknu.applyserver.auth.data.entity.GoogleProfile;
import com.likelionknu.applyserver.auth.data.entity.Profile;
import com.likelionknu.applyserver.auth.data.entity.User;
import com.likelionknu.applyserver.auth.data.enums.Role;
import com.likelionknu.applyserver.auth.data.repository.UserRepository;
import com.likelionknu.applyserver.auth.exception.GoogleAuthenticaionFailedException;
import com.likelionknu.applyserver.common.security.AuthenticationToken;
import com.likelionknu.applyserver.common.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.client.RestTemplate;

import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.util.*;

@Service
@Slf4j
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final JwtTokenProvider jwtTokenProvider;

    @Value("${google.client.id}")
    private String clientId;

    @Value("${google.client.secret}")
    private String clientSecret;

    @Value("${google.redirect.uri}")
    private String redirectUri;

    private GoogleProfile getGoogleProfile(String code) {
        try {
            RestTemplate restTemplate = new RestTemplate();

            String tokenUrl = "https://oauth2.googleapis.com/token";
            Map<String, String> params = new HashMap<>();
            params.put("code", code);
            params.put("client_id", clientId);
            params.put("client_secret", clientSecret);
            params.put("redirect_uri", redirectUri);
            params.put("grant_type", "authorization_code");

            ResponseEntity<Map<String, Object>> tokenResponse = restTemplate.exchange(
                    tokenUrl, HttpMethod.POST, new HttpEntity<>(params), new ParameterizedTypeReference<>() {}
            );

            Map<String, Object> tokenBody = tokenResponse.getBody();
            if (tokenBody == null || !tokenBody.containsKey("access_token")) {
                log.error("[getGoogleProfile] Google Access Token 발급 실패");
                throw new GoogleAuthenticaionFailedException();
            }
            String accessToken = tokenBody.get("access_token").toString();

            String userInfoUrl = "https://www.googleapis.com/oauth2/v2/userinfo";
            HttpHeaders headers = new HttpHeaders();
            headers.setBearerAuth(accessToken);
            HttpEntity<String> entity = new HttpEntity<>(headers);

            ResponseEntity<Map<String, Object>> userResponse = restTemplate.exchange(
                    userInfoUrl, HttpMethod.GET, entity, new ParameterizedTypeReference<>() {}
            );

            Map<String, Object> attributes = userResponse.getBody();
            if (attributes == null) {
                log.error("[getGoogleProfile] Google 소셜 프로필 조회 실패");
                throw new GoogleAuthenticaionFailedException();
            }

            return GoogleProfile.builder()
                    .email((String) attributes.get("email"))
                    .name((String) attributes.get("name"))
                    .profileUrl((String) attributes.get("picture"))
                    .build();
        } catch (HttpClientErrorException | HttpServerErrorException e) {
            throw new GoogleAuthenticaionFailedException();
        }
    }

    private Authentication createAuthentication(User user) {
        List<SimpleGrantedAuthority> authorities = Collections.singletonList(
                new SimpleGrantedAuthority(user.getRole().toString())
        );

        return new UsernamePasswordAuthenticationToken(user.getEmail(), null, authorities);
    }

    private User registerNewUser(GoogleProfile googleProfile) {
        User newUser = new User();
        newUser.setEmail(googleProfile.getEmail());
        newUser.setName(googleProfile.getName());
        newUser.setProfileUrl(googleProfile.getProfileUrl());
        newUser.setRole(Role.USER);
        newUser.setLastAccessAt(LocalDateTime.now());

        Profile profile = Profile.builder()
                .user(newUser)
                .build();

        newUser.setProfile(profile);

        userRepository.save(newUser);

        return newUser;
    }

    public TokenResponseDto userSocialSignIn(String code) {
        GoogleProfile googleProfile = getGoogleProfile(code);
        User user = userRepository.findByEmail(googleProfile.getEmail());

        if(user == null) {
            log.info("[userSocialSignIn] 새로운 사용자: {}", googleProfile.getEmail());
            user = registerNewUser(googleProfile);
        } else {
            log.info("[userSocialSignIn] 기존 가입된 사용자: {}", googleProfile.getEmail());
        }

        Authentication authentication = createAuthentication(user);
        AuthenticationToken authenticationToken = jwtTokenProvider.generateToken(authentication);
        return TokenResponseDto.builder()
                .accessToken(authenticationToken.getAccessToken())
                .refreshToken(authenticationToken.getRefreshToken())
                .name(user.getName())
                .role(user.getRole().toString())
                .build();
    }

    public void userLogout(String email) {
        log.info("[userLogout] 사용자 로그아웃: {}", email);
        jwtTokenProvider.expireUsersToken(email);
    }

    public TokenResponseDto reissue(String refreshToken) {
        jwtTokenProvider.validateRefreshToken(refreshToken);

        String email = jwtTokenProvider.getSubject(refreshToken);
        User user = userRepository.findByEmail(email);
        Authentication authentication = createAuthentication(user);
        AuthenticationToken authenticationToken = jwtTokenProvider.generateToken(authentication);
        log.info("[reissue] 사용자 JWT 재발급 완료: {}", user.getEmail());

        return TokenResponseDto.builder()
                .accessToken(authenticationToken.getAccessToken())
                .refreshToken(authenticationToken.getRefreshToken())
                .name(user.getName())
                .role(user.getRole().toString())
                .build();
    }
}

File: src/main/kotlin/com/likelionknu/applyserver/auth/service/AuthService.kt
package com.likelionknu.applyserver.auth.service

import com.likelionknu.applyserver.auth.data.dto.response.TokenResponseDto
import com.likelionknu.applyserver.auth.data.entity.GoogleProfile
import com.likelionknu.applyserver.auth.data.entity.Profile
import com.likelionknu.applyserver.auth.data.entity.User
import com.likelionknu.applyserver.auth.data.enums.PlatformDivider
import com.likelionknu.applyserver.auth.data.enums.Role
import com.likelionknu.applyserver.auth.exception.GoogleAuthenticaionFailedException
import com.likelionknu.applyserver.auth.repository.UserRepository
import com.likelionknu.applyserver.common.security.JwtTokenProvider
import com.likelionknu.applyserver.discord.service.DiscordNotificationService
import com.likelionknu.applyserver.mail.data.dto.MailRequestDto
import com.likelionknu.applyserver.mail.data.entity.MailContent
import com.likelionknu.applyserver.mail.service.MailService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Service
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.HttpServerErrorException
import org.springframework.web.client.RestTemplate
import java.time.LocalDateTime

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val jwtTokenProvider: JwtTokenProvider,
    private val mailService: MailService,
    private val discordNotificationService: DiscordNotificationService,
    @Value("\${google.client.id}") private val clientId: String,
    @Value("\${google.client.secret}") private val clientSecret: String,
    @Value("\${google.redirect.uri}") private val redirectUri: String,
    @Value("\${google.redirect.admin}") private val redirectUriAdmin: String
) {
    private val log = LoggerFactory.getLogger(AuthService::class.java)

    private fun getGoogleProfile(code: String, platformDivider: PlatformDivider): GoogleProfile {
        return try {
            val restTemplate = RestTemplate()

            val redirect = when (platformDivider) {
                PlatformDivider.APPLY -> {
                    log.info("[getGoogleProfile] 어플라이 소셜 로그인 시도")
                    redirectUri
                }
                PlatformDivider.ADMIN -> {
                    log.info("[getGoogleProfile] 관리자 소셜 로그인 시도")
                    redirectUriAdmin
                }
                else -> redirectUri
            }

            val tokenUrl = "https://oauth2.googleapis.com/token"
            val params = hashMapOf(
                "code" to code,
                "client_id" to clientId,
                "client_secret" to clientSecret,
                "redirect_uri" to redirect,
                "grant_type" to "authorization_code"
            )

            val tokenResponse: ResponseEntity<Map<String, Any>> = restTemplate.exchange(
                tokenUrl,
                HttpMethod.POST,
                HttpEntity(params),
                object : ParameterizedTypeReference<Map<String, Any>>() {}
            )

            val tokenBody = tokenResponse.body
            if (tokenBody == null || !tokenBody.containsKey("access_token")) {
                log.error("[getGoogleProfile] Google Access Token 발급 실패")
                throw GoogleAuthenticaionFailedException()
            }

            val accessToken = tokenBody["access_token"].toString()

            val userInfoUrl = "https://www.googleapis.com/oauth2/v2/userinfo"
            val headers = HttpHeaders().apply {
                setBearerAuth(accessToken)
            }

            val userResponse: ResponseEntity<Map<String, Any>> = restTemplate.exchange(
                userInfoUrl,
                HttpMethod.GET,
                HttpEntity<String>(headers),
                object : ParameterizedTypeReference<Map<String, Any>>() {}
            )

            val attributes = userResponse.body
            if (attributes == null) {
                log.error("[getGoogleProfile] Google 소셜 프로필 조회 실패")
                throw GoogleAuthenticaionFailedException()
            }

            GoogleProfile(
                email = attributes["email"] as String,
                name = attributes["name"] as String,
                profileUrl = attributes["picture"] as String
            )
        } catch (e: HttpClientErrorException) {
            throw GoogleAuthenticaionFailedException()
        } catch (e: HttpServerErrorException) {
            throw GoogleAuthenticaionFailedException()
        }
    }

    private fun createAuthentication(user: User): Authentication {
        val authorities = listOf(
            SimpleGrantedAuthority("ROLE_${user.role}")
        )

        return UsernamePasswordAuthenticationToken(user.email, null, authorities)
    }

    private fun registerNewUser(googleProfile: GoogleProfile): User {
        val newUser = User().apply {
            email = googleProfile.email
            name = googleProfile.name
            profileUrl = googleProfile.profileUrl
            role = Role.USER
            modifiedAt = LocalDateTime.now()
            registeredAt = LocalDateTime.now()
            lastAccessAt = LocalDateTime.now()
        }

        val profile = Profile().apply {
            user = newUser
        }

        newUser.profile = profile

        userRepository.save(newUser)

        discordNotificationService.sendNewUserNotification(
            newUser.name,
            newUser.email,
            newUser.profileUrl
        )

        return newUser
    }

    fun userSocialSignIn(code: String, platformDivider: PlatformDivider): TokenResponseDto {
        val googleProfile = getGoogleProfile(code, platformDivider)
        var user = userRepository.findOptionalByEmail(googleProfile.email).orElse(null)
        var isNewUser = false

        if (user == null) {
            log.info("[userSocialSignIn] 새로운 사용자: {}", googleProfile.email)
            user = registerNewUser(googleProfile)

            val mailContentList = listOf(
                MailContent(
                    key = "userName",
                    value = user.name
                )
            )

            mailService.sendMail(
                MailRequestDto(
                    user = null,
                    email = user.email,
                    title = "회원가입 완료 안내",
                    template = "register-success",
                    dataList = mailContentList
                )
            )

            isNewUser = true
        } else {
            log.info("[userSocialSignIn] 기존 가입된 사용자: {}", googleProfile.email)
        }

        user.lastAccessAt = LocalDateTime.now()
        userRepository.save(user)

        val authentication = createAuthentication(user)
        val authenticationToken = jwtTokenProvider.generateToken(authentication)

        return TokenResponseDto(
            accessToken = authenticationToken.accessToken,
            refreshToken = authenticationToken.refreshToken,
            name = user.name,
            role = user.role.toString(),
            isNewUser = isNewUser
        )
    }

    fun userLogout(email: String) {
        log.info("[userLogout] 사용자 로그아웃: {}", email)
        jwtTokenProvider.expireUsersToken(email)
    }

    fun reissue(refreshToken: String): TokenResponseDto {
        jwtTokenProvider.validateRefreshToken(refreshToken)

        val email = jwtTokenProvider.getSubject(refreshToken)
        val user = userRepository.findOptionalByEmail(email)
            .orElseThrow { GoogleAuthenticaionFailedException() }

        val authentication = createAuthentication(user)
        val authenticationToken = jwtTokenProvider.generateToken(authentication)

        log.info("[reissue] 사용자 JWT 재발급 완료: {}", user.email)

        return TokenResponseDto(
            accessToken = authenticationToken.accessToken,
            refreshToken = authenticationToken.refreshToken,
            name = user.name,
            role = user.role.toString()
        )
    }
}